[quant][graphmode][fx] Support inplace option

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43984 [quant][graphmode][fx][api] Call fuse in prepare
* **#43983 [quant][graphmode][fx] Support inplace option**
* #44042 [quant][graphmode][fx] Support quantize per channel in all cases

Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

jerryzh168 committed:
[quant][graphmode][fx] Support inplace option

Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

vkuzo commented on torch/quantization/quantize_fx.py:
```diff
@@ -91,22 +91,18 @@ def _quantize_fx(model, qconfig_dict, run_fn=None, run_args=None, inplace=False,
-        # TODO: change inplace to True since the model is already copied in
-        # prepare
-        model = convert_dynamic_fx(model, False, debug)
+        model = convert_dynamic_fx(model, inplace=True, debug=debug)
```
should this be `inplace=inplace`?

jerryzh168 replied:
this should be True since the inplace option is already applied in the previous step(prepare)

vkuzo commented on torch/quantization/quantize_fx.py:
```diff
@@ -91,22 +91,18 @@ def _quantize_fx(model, qconfig_dict, run_fn=None, run_args=None, inplace=False,
-        # TODO: change inplace to True since the model is already copied in
-        # prepare
-        model = convert_fx(model, False, debug)
+        model = convert_fx(model, inplace=True, debug=debug)
```
and here

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -282,26 +281,24 @@ def _run_weight_observers(self, observed):
-        self.modules = dict(observed_root.named_modules())
+        model.eval().cpu()
+        model_root = model.root
+        model_graph = model.graph
```
optional nit: can we just use `model.graph`?

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -514,10 +512,11 @@ def load_arg(a):
 
-    def convert(self, observed, inplace=False, debug=False, is_dynamic=False):
-        quantized = self._convert(observed, inplace, debug, is_dynamic)
+    def convert(self, model, inplace=False, debug=False, is_dynamic=False):
```
thanks for renaming, the new name is much clearer

vkuzo reviewed (approved):
lg, seems like the `inplace=inplace` should be fixed before landing

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support inplace option"


Summary:
Support inplace option in apis

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23459260](https://our.internmc.facebook.com/intern/diff/D23459260)

[ghstack-poisoned]
